<h1>This Notebook's purpose is to create an analysis of the music that I listen to using my spotify and predict a minimum of 5 new songs that I may be interested in listening to based on the songs that exist in the playlist.</h1>

<h2>Importing required packages</h2>

In [195]:
import pandas as pd #Used for dataset structure
import matplotlib.pyplot as plt
%matplotlib inline

In [196]:
df = pd.read_csv('Local Files.csv') #Reading in my dataset
pd.options.display.max_columns = 50 #Force pandas to show me all columns

<h2>Data Exploration</h2>

In [197]:
df.head() #Shows first 5 rows of the dataset

added_at        added_by-external_urls-spotify  \
0  2017-05-14T23:11:48Z  https://open.spotify.com/user/bauzor   
1  2017-05-14T23:11:48Z  https://open.spotify.com/user/bauzor   
2  2017-05-14T23:11:48Z  https://open.spotify.com/user/bauzor   
3  2017-05-14T23:11:48Z  https://open.spotify.com/user/bauzor   
4  2017-05-14T23:11:48Z  https://open.spotify.com/user/bauzor   

                             added_by-href added_by-id added_by-type  \
0  https://api.spotify.com/v1/users/bauzor      bauzor          user   
1  https://api.spotify.com/v1/users/bauzor      bauzor          user   
2  https://api.spotify.com/v1/users/bauzor      bauzor          user   
3  https://api.spotify.com/v1/users/bauzor      bauzor          user   
4  https://api.spotify.com/v1/users/bauzor      bauzor          user   

          added_by-uri  is_local  primary_color track-album-album_type  \
0  spotify:user:bauzor     False            NaN                  album   
1  spotify:user:bauzor     False            NaN                  album   
2  spotify:user:bauzor      True            NaN                    NaN   
3  spotify:user:bauzor      True            NaN                    NaN   
4  spotify:user:bauzor      True            NaN                    NaN   

           track-album-artists-external_urls-spotify  \
0  https://open.spotify.com/artist/7oPftvlwr6VrsV...   
1  https://open.spotify.com/artist/7oPftvlwr6VrsV...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                            track-album-artists-href  track-album-artists-id  \
0  https://api.spotify.com/v1/artists/7oPftvlwr6V...  7oPftvlwr6VrsViSDV7fJY   
1  https://api.spotify.com/v1/artists/7oPftvlwr6V...  7oPftvlwr6VrsViSDV7fJY   
2                                                NaN                     NaN   
3                                                NaN                     NaN   
4                                                NaN                     NaN   

  track-album-artists-name track-album-artists-type  \
0                Green Day                   artist   
1                Green Day                   artist   
2                      NaN                      NaN   
3                      NaN                      NaN   
4                      NaN                      NaN   

                 track-album-artists-uri  \
0  spotify:artist:7oPftvlwr6VrsViSDV7fJY   
1  spotify:artist:7oPftvlwr6VrsViSDV7fJY   
2                                    NaN   
3                                    NaN   
4                                    NaN   

                   track-album-external_urls-spotify  \
0  https://open.spotify.com/album/4UV0uKJYgaYeHu9...   
1  https://open.spotify.com/album/4UV0uKJYgaYeHu9...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                    track-album-href          track-album-id  \
0  https://api.spotify.com/v1/albums/4UV0uKJYgaYe...  4UV0uKJYgaYeHu9WK1KfC0   
1  https://api.spotify.com/v1/albums/4UV0uKJYgaYe...  4UV0uKJYgaYeHu9WK1KfC0   
2                                                NaN                     NaN   
3                                                NaN                     NaN   
4                                                NaN                     NaN   

   track-album-images-height  \
0                       64.0   
1                       64.0   
2                        NaN   
3                        NaN   
4                        NaN   

                              track-album-images-url  \
0  https://i.scdn.co/image/c56af2618d8d9d8680c4b7...   
1  https://i.scdn.co/image/c56af2618d8d9d8680c4b7...   
2                                                NaN   
3                                                NaN   
4                                        

<p>Can't see all 50 columns in our dataset clearly so let's use .info() to give us the column headers as well null value info</p>

In [198]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4598 entries, 0 to 4597
Data columns (total 50 columns):
added_at                                     4598 non-null object
added_by-external_urls-spotify               4598 non-null object
added_by-href                                4598 non-null object
added_by-id                                  4598 non-null object
added_by-type                                4598 non-null object
added_by-uri                                 4598 non-null object
is_local                                     4598 non-null bool
primary_color                                0 non-null float64
track-album-album_type                       2107 non-null object
track-album-artists-external_urls-spotify    2107 non-null object
track-album-artists-href                     2107 non-null object
track-album-artists-id                       2107 non-null object
track-album-artists-name                     2104 non-null object
track-album-artists-type               

In [199]:
df['added_by-external_urls-spotify'].value_counts()

https://open.spotify.com/user/bauzor    4598
Name: added_by-external_urls-spotify, dtype: int64

In [200]:
df[['added_by-href','added_by-type','added_by-id','added_by-uri','added_by-uri']].duplicated().value_counts()

True     4597
False       1
dtype: int64

<p>Notice the fact that there are 4598 entries given by the RangeIndex from .info() and for all 4598 values we get that the user bauzor added every song to this playlist, thus the columns 'added_by-XXXXXX' are really useless since they are all the same and feeding it to a model would confuse the model thinking that these entries are all correlated thus it'd be best to drop these columns.</p>

In [201]:
df = df.drop(['added_by-external_urls-spotify','added_by-href','added_by-type','added_by-id','added_by-uri','added_by-uri'], axis=1)
df

added_at  is_local  primary_color track-album-album_type  \
0     2017-05-14T23:11:48Z     False            NaN                  album   
1     2017-05-14T23:11:48Z     False            NaN                  album   
2     2017-05-14T23:11:48Z      True            NaN                    NaN   
3     2017-05-14T23:11:48Z      True            NaN                    NaN   
4     2017-05-14T23:11:48Z      True            NaN                    NaN   
5     2017-05-14T23:11:48Z      True            NaN                    NaN   
6     2017-05-14T23:11:48Z      True            NaN                    NaN   
7     2017-05-14T23:11:48Z      True            NaN                    NaN   
8     2017-05-14T23:11:48Z      True            NaN                    NaN   
9     2017-05-14T23:11:48Z      True            NaN                    NaN   
10    2017-05-14T23:11:48Z      True            NaN                    NaN   
11    2017-05-14T23:11:48Z     False            NaN                  album   
12    2017-05-14T23:11:48Z      True            NaN                    NaN   
13    2017-05-14T23:11:48Z      True            NaN                    NaN   
14    2017-05-14T23:11:48Z      True            NaN                    NaN   
15    2017-05-14T23:11:48Z      True            NaN                    NaN   
16    2017-05-14T23:11:48Z      True            NaN                    NaN   
17    2017-05-14T23:11:48Z      True            NaN                    NaN   
18    2017-05-14T23:11:48Z      True            NaN                    NaN   
19    2017-05-14T23:11:48Z      True            NaN                    NaN   
20    2017-05-14T23:11:48Z      True            NaN                    NaN   
21    2017-05-14T23:11:48Z     False            NaN                  album   
22    2017-05-14T23:11:48Z      True            NaN                    NaN   
23    2017-05-14T23:11:48Z      True            NaN                    NaN   
24    2017-05-14T23:11:48Z      True            NaN                    NaN   
25    2017-05-14T23:11:48Z      True            NaN                    NaN   
26    2017-05-14T23:11:48Z      True            NaN                    NaN   
27    2017-05-14T23:11:48Z      True            NaN                    NaN   
28    2017-05-14T23:11:48Z      True            NaN                    NaN   
29    2017-05-14T23:11:48Z     False            NaN                  album   
...                    ...       ...            ...                    ...   
4568  2018-10-06T18:47:59Z      True            NaN                    NaN   
4569  2018-10-06T18:47:59Z      True            NaN                    NaN   
4570  2018-10-06T18:47:59Z      True            NaN                    NaN   
4571  2018-10-06T18:47:59Z      True            NaN                    NaN   
4572  2018-10-06T18:47:59Z      True            NaN                    NaN   
4573  2018-10-06T18:47:59Z      True            NaN                    NaN   
4574  2018-10-06T18:47:59Z      True            NaN                    NaN   
4575  2018-10-09T18:15:07Z     False            NaN                 single   
4576  2018-10-12T16:07:10Z     False            NaN                 single   
4577  2018-10-12T16:35:09Z     False            NaN                 single   
4578  2018-10-15T15:51:45Z     False            NaN                 single   
4579  2018-10-15T15:51:49Z     False            NaN                 single   
4580  2018-10-15T16:08:18Z     False            NaN                  album   
4581  2018-10-15T17:37:26Z     False            NaN                  album   
4582  2018-10-15T17:37:36Z     False            NaN                  album   
4583  2018-10-15T18:02:21Z     False            NaN                  album   
4584  2018-10-15T23:24:50Z     False            NaN                  album   
4585  2018-10-19T01:29:19Z     False            NaN                 single   
4586  2018-10-19T02:07:14Z     False            NaN                 single   
4587  2018-10-19T15:50:09Z     False    

In [202]:
df['primary_color'].value_counts(dropna=False)

NaN    4598
Name: primary_color, dtype: int64

<p>Notice that all values of primary color are unknown, thus we also drop this column</p>

In [203]:
df = df.drop(['primary_color'], axis=1)
df

added_at  is_local track-album-album_type  \
0     2017-05-14T23:11:48Z     False                  album   
1     2017-05-14T23:11:48Z     False                  album   
2     2017-05-14T23:11:48Z      True                    NaN   
3     2017-05-14T23:11:48Z      True                    NaN   
4     2017-05-14T23:11:48Z      True                    NaN   
5     2017-05-14T23:11:48Z      True                    NaN   
6     2017-05-14T23:11:48Z      True                    NaN   
7     2017-05-14T23:11:48Z      True                    NaN   
8     2017-05-14T23:11:48Z      True                    NaN   
9     2017-05-14T23:11:48Z      True                    NaN   
10    2017-05-14T23:11:48Z      True                    NaN   
11    2017-05-14T23:11:48Z     False                  album   
12    2017-05-14T23:11:48Z      True                    NaN   
13    2017-05-14T23:11:48Z      True                    NaN   
14    2017-05-14T23:11:48Z      True                    NaN   
15    2017-05-14T23:11:48Z      True                    NaN   
16    2017-05-14T23:11:48Z      True                    NaN   
17    2017-05-14T23:11:48Z      True                    NaN   
18    2017-05-14T23:11:48Z      True                    NaN   
19    2017-05-14T23:11:48Z      True                    NaN   
20    2017-05-14T23:11:48Z      True                    NaN   
21    2017-05-14T23:11:48Z     False                  album   
22    2017-05-14T23:11:48Z      True                    NaN   
23    2017-05-14T23:11:48Z      True                    NaN   
24    2017-05-14T23:11:48Z      True                    NaN   
25    2017-05-14T23:11:48Z      True                    NaN   
26    2017-05-14T23:11:48Z      True                    NaN   
27    2017-05-14T23:11:48Z      True                    NaN   
28    2017-05-14T23:11:48Z      True                    NaN   
29    2017-05-14T23:11:48Z     False                  album   
...                    ...       ...                    ...   
4568  2018-10-06T18:47:59Z      True                    NaN   
4569  2018-10-06T18:47:59Z      True                    NaN   
4570  2018-10-06T18:47:59Z      True                    NaN   
4571  2018-10-06T18:47:59Z      True                    NaN   
4572  2018-10-06T18:47:59Z      True                    NaN   
4573  2018-10-06T18:47:59Z      True                    NaN   
4574  2018-10-06T18:47:59Z      True                    NaN   
4575  2018-10-09T18:15:07Z     False                 single   
4576  2018-10-12T16:07:10Z     False                 single   
4577  2018-10-12T16:35:09Z     False                 single   
4578  2018-10-15T15:51:45Z     False                 single   
4579  2018-10-15T15:51:49Z     False                 single   
4580  2018-10-15T16:08:18Z     False                  album   
4581  2018-10-15T17:37:26Z     False                  album   
4582  2018-10-15T17:37:36Z     False                  album   
4583  2018-10-15T18:02:21Z     False                  album   
4584  2018-10-15T23:24:50Z     False                  album   
4585  2018-10-19T01:29:19Z     False                 single   
4586  2018-10-19T02:07:14Z     False                 single   
4587  2018-10-19T15:50:09Z     False                 single   
4588  2018-10-21T02:45:58Z     False                  album   
4589  2018-10-21T02:48:33Z      True                    NaN   
4590  2018-10-21T03:27:35Z     False                  album   
4591  2018-10-21T03:31:55Z     False                  album   
4592  2018-10-21T03:57:01Z     False                 single   
4593  2018-10-21T04:03:37Z     False                  album   
4594  2018-10-21T04:26:32Z     False                  album   
4595  2018-10-22T23:29:10Z     False                 single   
4596  2018-10-23T19:50:15Z     False                  album   
4597  2018-10-23T20:08:51Z     False                  album   

              track-album-artists-external_urls-spotify  \
0     https://open.spotify.com/artist/7oPftvlwr6VrsV

<p>There are still a bunch of random columns that I believe won't help us engineer any new features and will not contribute to the rest of the analysis at all in terms of information provided so I have listed all the columns as well as why I do not think they are necessary and appended it to one list to drop all at once at the very end.</p>

In [204]:
df['track-album-album_type'].value_counts(dropna=False) 
#no value coming out of track-album-album_type because it just tells us whether a song is from an album or not and
#this provides me zero value in whether or not I would enjoy a new song based on if it was packaged in an album or not.

NaN            2491
album          1609
single          373
compilation     125
Name: track-album-album_type, dtype: int64

In [205]:
drop=[]
drop.append('track-album-album_type')
drop

['track-album-album_type']

<p>I'm going to keep both the external_urls-spotify as well as the href in case I might need them later to look for new songs from the same artists. There are also some songs that are local files that need to be cleaned up to have their artist links specified since some of the songs that I have are on spotify but I have the local file version of the song. For track-album-artists- external_urls-spotify, href, id and name I think they might be useful for comparing with other people's playlists because normally if you like one song from a particulary artist, you might like other songs that are from that artist as well as maybe finding the genre a particulary artist is.</p>

In [206]:
df['track-album-artists-type'].value_counts(dropna=False)

NaN       2491
artist    2107
Name: track-album-artists-type, dtype: int64

<p>Doesn't seem like I can do much with these values still since it provides no value</p>

In [207]:
drop.append('track-album-artists-type')
drop

['track-album-album_type', 'track-album-artists-type']

<p>Next we have track-album-artists-uri, external_urls-spotify, href, id. If a user has 10 out of the 11 songs of an album saved, recommending the 11th song in the album might be a good idea thus I have kept these columns for now until I decide how and what kind of recommendation model I'll be using</p>

<p>Judging based on the column header we have the albums image height width and url. I personally believe that album art does not have a huge correlation to whether or not songs from an album would be enjoyed by an individual thus I will be making the choice to drop it from predictions.</p>

In [208]:
drop.extend(('track-album-images-height', 'track-album-images-url', 'track-album-images-width'))
drop

['track-album-album_type',
 'track-album-artists-type',
 'track-album-images-height',
 'track-album-images-url',
 'track-album-images-width']

<p>Next we have the (track-album-) release_date, release_date_precision and the total_tracks, I feel that these columns are also quite useless and have no application to whether or not a user would enjoy a new song based on albums release date thus I have also added this to the drop list.</p> 

In [209]:
drop.extend(('track-album-release_date','track-album-release_date_precision','track-album-total_tracks'))
drop

['track-album-album_type',
 'track-album-artists-type',
 'track-album-images-height',
 'track-album-images-url',
 'track-album-images-width',
 'track-album-release_date',
 'track-album-release_date_precision',
 'track-album-total_tracks']

In [210]:
df['track-album-release_date_precision'].value_counts(dropna=False)

NaN     2491
day     2048
year      59
Name: track-album-release_date_precision, dtype: int64

<p>track-album-type tells us nothing so we will drop it as well</p>

In [211]:
df['track-album-type'].value_counts(dropna=False)

album    4598
Name: track-album-type, dtype: int64

In [212]:
drop.append('track-album-type')
drop